### ISSUE TITLE
Start finetuning self-recognition for our best models

### ISSUE THREAD

**Issue by NelsonG-C** [NelsonG-C/lasr-labs-2025-control-project#95]
TBC on details after meeting with Charlie

**Comment by NelsonG-C** [comment]
@Barahlush Make sure to note down your process as you go so we can easily do a writeup of what finetuning we did

**Comment by Barahlush** [comment]
Started with setting up OpenAI fine-tuning, using web-interface. 
Initial goal was to see if fine-tuning will improve accuracy of a small model (gpt-4.1-mini) on individual self-rec. 

Generated a bunch of data from the APPS training set:
- 9 solutions by different models (`['gpt-5-nano', 'human', 'gpt-4.1-nano', 'gemini-2.5-flash', 'o4-mini', 'gemini-2.5-flash-lite', 'gpt-4o-mini', 'claude-3-5-haiku-20241022', 'gpt-4.1-mini']`) for the first 15 problems (total 120, 15 of them are by gpt-4.1-mini)
- 1 solution by gpt-4.1-mini for each of the problems 16-120 (total 105, to balance the dataset).

I created an eval using all solutions problems 14-15 (18 solutions, 2 from gpt-4.1-mini) and 107-120 (14 solutions, all from gpt-4.1-mini) with total of *32 solutions*, 16 from gpt-4.1-mini.

For the training I used all remaining problems, and I split them into train/val by shuffling and splitting 188/20.

So, all the splits train/val/eval do not intersect.

Training run: https://platform.openai.com/finetune/ftjob-f2Zuc6n4UBgp6JZ2zSEqz74j?filter=all
Evals: https://platform.openai.com/evaluations/eval_68b7161ed89881918c0f0cbfb44d8458

**Training improved accuracy on eval from 56% to 62%.**
I obtained 3 checkpoints (1 final and 2 mid-training), first shows same accuracy as the base model - 56%, the last two show 62%.

I tried to improve accuracy of the base model using prompts (e.g. asking for reasoning or confidence score, providing model name) - didn't achieve any improvement over the simple prompt, **best accuracy of prompt-engineering was 56%**, same as the base model.

There are a couple of direction to develop that:
1) generate more data for evals and training (32 for evals seems too low to detect small changes)
2) use bigger models, gpt-4.1-mini is pretty weak generally (but need to regenerate all the datasets)

**Comment by Barahlush** [comment]
Morgan proposed to do these small fine-tunings on all models and select by post-finetuning accuracy, and then run finetuning of the chosen model on more data. That makes sense, anyways, I need to write some code to carefully and automatically build balanced datasets for any model I select. I'll work on this today.

**Comment by Barahlush** [comment]
Okay, it seems OpenAI only allows to fine-tune a limited set of models:
gpt-4.1-nano-2025-04-14
gpt-4.1-mini-2025-04-14
gpt-4.1-2025-04-14
gpt-4o-mini-2024-07-18
gpt-4o-2024-08-06
gpt-3.5-turbo-1106
gpt-3.5-turbo-0125

I'm planning to eval the big ones: gpt-4.1-2025-04-14 and gpt-4o-2024-08-06 and fine-tune the one of them which will show best on eval

Also, found out that we can fine-tune Gemini models: https://console.cloud.google.com/vertex-ai/studio/tuning/custom?project=gen-lang-client-0787837331

Plan to look into that this today-tomorrow.

**Comment by Barahlush** [comment]
Finetuned the `gpt-4.1-2025-04-14`, since it provided better results than `gpt-4o-2024-08-06` on eval.

## Results
**Target Model:** `ft:gpt-4.1-2025-04-14:lasr-labs::CC5IeIFO`
**Total Evaluation Records:** 120
**Overall Accuracy:** 0.9417
**Overall Precision:** 0.9048
**Overall Recall:** 0.7917
**Overall F1 Score:** 0.8444
**Confusion Matrix:**
  _True Negatives:_  94
  _False Positives:_ 2
  _False Negatives:_ 5
  _True Positives:_  19


## Setup
I generated a bunch of data (120 x 5) using the following models: 
```
             gpt-4o-2024-08-06:   120 solutions
      claude-sonnet-4-20250514:   120 solutions
gemini-2.5-flash-preview-05-20:   120 solutions
            gpt-4.1-2025-04-14:   120 solutions
                         human:   120 solutions
```
Each model generated exactly 1 solution per problem, so there are 120 problems in the dataset (first 120 problems of APPS train split, filtered).

I split the data into train and test as 96/24 (80%/20%), so first 96 problems are used for training, others are used for evaluation (including the comparison between 4.1 and 4o).

To build a training dataset, I took 96 solutions from each model, and to balance the dataset (96 vs 384) I upsample solutions from gpt-4.1 (just by duplicating them x4), so the final training dataset contains 768 solutions. Dataset is shuffled before training.

**I use eval set as is, imbalanced 1:4, so accuracy is not the best metric here.** _Precision/recall are more informative._

## Concerns and ideas
- Training costed $58 (without train/eval data generation), 4.1 is pretty expensive.
- Finetuning can influence the model's generated solutions distribution. We should probably check self rec quality on the solutions of the finetuned model, it can actually be lower than reported because of that. OOH, we can use the finetuned model only as a monitor, and as a generator use the base one (though it's not the cleanest control setup in our scenario).
- Interesting to see if finetuning on code self-rec improved self-rec on non-code texts. Code is pretty specific, so may be big if yes (the model learns to use the dark matter between the tokens)
